gst/nnstreamer/nnstreamer_watchdog.c

```diff
@@ -0,0 +1,108 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * NNStreamer watchdog
+ * Copyright (C) 2024 Gichan Jang <gichan2.jang@samsung.com>
+ */
+/**
+ * @file	nnstreamer_watchdog.c
+ * @date	30 Oct 2024
+ * @brief	NNStreamer watchdog to manage the schedule in the element.
+ * @see		https://github.com/nnstreamer/nnstreamer
+ * @author	Gichan Jang <gichan2.jang@samsung.com>
+ * @bug		No known bugs except for NYI items
+ */
+
+#include "nnstreamer_watchdog.h"
+
+/**
+ * @brief Structure for NNStreamer watchdog.
+ */
+typedef struct _NnstWatchdog
+{
+  GMainContext *main_context;
+  GMainLoop *main_loop;
+  GThread *thread;
+  GSource *source;
+} NnstWatchdog;
+
+/**
+ * @brief Watchdog thread.
+ */
+static gpointer
+_nnstreamer_watchdog_thread (gpointer ptr)
+{
+  NnstWatchdog *watchdog = (NnstWatchdog *) ptr;
+
+  g_main_loop_run (watchdog->main_loop);
+
+  return NULL;
+}
+
+/**
+ * @brief Create nnstreamer watchdog. Recommended using watchdog handle with proper lock (e.g., GST_OBJECT_LOCK())
+ */
+void
+nnstreamer_watchdog_create (nns_watchdog_h * watchdog_h)
+{
+  NnstWatchdog *watchdog = g_new0 (NnstWatchdog, 1);
+
+  watchdog->main_context = g_main_context_new ();
+  watchdog->main_loop = g_main_loop_new (watchdog->main_context, FALSE);
+  watchdog->thread =
+      g_thread_new ("suspend_watchdog", _nnstreamer_watchdog_thread, watchdog);
+
+  *watchdog_h = watchdog;
+}
+
+/**
+ * @brief Destroy watchdog source. Recommended using watchdog handle with proper lock (e.g., GST_OBJECT_LOCK())
+ */
+void
+nnstreamer_watchdog_destroy (nns_watchdog_h watchdog_h)
+{
+  NnstWatchdog *watchdog = (NnstWatchdog *) watchdog_h;
+  nnstreamer_watchdog_release (watchdog);
+
+  if (watchdog && watchdog->main_context) {
+    g_main_loop_quit (watchdog->main_loop);
+    g_thread_join (watchdog->thread);
+    watchdog->thread = NULL;
+
+    g_main_loop_unref (watchdog->main_loop);
+    watchdog->main_loop = NULL;
+
+    g_main_context_unref (watchdog->main_context);
+    watchdog->main_context = NULL;
+
+    g_free (watchdog_h);
+  }
+}
+
+/**
+ * @brief Release watchdog source. Recommended using watchdog handle with proper lock (e.g., GST_OBJECT_LOCK())
+ */
+void
+nnstreamer_watchdog_release (nns_watchdog_h watchdog_h)
+{
+  NnstWatchdog *watchdog = (NnstWatchdog *) watchdog_h;
+  if (watchdog && watchdog->source) {
+    g_source_destroy (watchdog->source);
+    g_source_unref (watchdog->source);
+    watchdog->source = NULL;
+  }
+}
+
+/**
+ * @brief Set watchdog timeout. Recommended using watchdog handle with proper lock (e.g., GST_OBJECT_LOCK())
+ */
+void
+nnstreamer_watchdog_feed (nns_watchdog_h watchdog_h, GSourceFunc func,
+    guint interval, void *user_data)
+{
+  NnstWatchdog *watchdog = (NnstWatchdog *) watchdog_h;
```

Comment:
A null check for watchdog is required.
if nnstreamer_watchdog_create() is failed, priv->watchdog_h is NULL.
```
[usecase]
if (priv->prop.suspend != 0) {
    GST_OBJECT_LOCK (self);
    nnstreamer_watchdog_create (&priv->watchdog_h);
    nnstreamer_watchdog_feed (priv->watchdog_h, gst_tensor_filter_watchdog_trigger, priv->prop.suspend, priv);
    GST_OBJECT_UNLOCK (self);
  }
```

Reply:
Changed to check NULL.